app/src/main/java/org/simple/clinic/patient/filter/SortByWeightedNameParts.kt

```diff
@@ -0,0 +1,54 @@
+package org.simple.clinic.patient.filter
+
+import kotlin.math.max
+
+/**
+ * This is a comparator that is used to sort [PatientSearchContext] objects by prioritizing matches
+ * which have the same number of words as the search term while still providing good
+ * results when the search term and the name do not have the same number of words.
+ *
+ * It works by generating an overall score for a match based on a weighted function that
+ * multiples the edit distance for a particular part with a multiplier which is decided
+ * based on the position of the term in the word.
+ *
+ * Variations in the earlier parts of the name are penalised more than the variations in the
+ * later parts of the name and names with more matching words get a better score.
+ **/
+class SortByWeightedNameParts : Comparator<PatientSearchContext> {
+
+  override fun compare(first: PatientSearchContext, second: PatientSearchContext): Int {
+    val scoreOfFirst = computeScore(first)
+    val scoreOfSecond = computeScore(second)
+
+    return scoreOfFirst.compareTo(scoreOfSecond)
+  }
+
+  private fun computeScore(searchInfo: PatientSearchContext): Double {
+    val maximumWordParts = max(searchInfo.nameParts.size, searchInfo.searchParts.size)
+
+    // Generates decreasing multipliers based on the position of the
+    // term. i.e, if the maximum number of parts is 3, the multipliers are
+    // 3, 2, 1 for positions 0, 1, 2 respectively.
+    val positionMultipliers = (1..maximumWordParts).map { it }.reversed()
+
+    // Generate weighted scores for each position by assigning an arbitrary value,
+    // one divided by the max parts in this case, and multiplying that value with
+    // the position multipliers generated previously.
+    val defaultScores = (1..maximumWordParts).map { (1 / maximumWordParts.toDouble()) * positionMultipliers[it - 1] }
+
+    // Generate the final score by multiplying the distance (normalized by dividing it by the
+    // total sum of the distances) with the weighted scores computed earlier and adding all the
+    // individual scores.
+    val sumOfDistances = searchInfo.totalEditDistance
+    val scores = defaultScores.toMutableList()
+
+    searchInfo.editDistances.forEachIndexed { index, (_, _, distance) ->
+      scores[index] *= when (sumOfDistances) {
+        0.0 -> 0.0
+        else -> distance / sumOfDistances
+      }
+    }
+
+    return scores.sum()
```

Comment:
Sanchita and I tried rewording things to make this block of code more readable on the lines of how you helped us understand the maths:

```kotlin
// Generate decreasing multipliers based on the position of the
// term. i.e, if the maximum number of parts is 3, the multipliers are
// 3, 2, 1 for positions 0, 1, 2 respectively.
val penaltiesForPositions: List<Int> = (1..maximumWordParts).map { it }.reversed()

// Generate weighted scores for each position by assigning a default value,
// one divided by the max parts in this case, and multiplying that value with
// the position multipliers generated previously.
val triedAndTestedArbitraryValue = 1 / maximumWordParts.toDouble()
val positionBasedDefaultWeights = penaltiesForPositions.map { penalty -> penalty * triedAndTestedArbitraryValue }

val editDistances = searchInfo.editDistances.map { it.editDistance }

return positionBasedDefaultWeights
    .zip(editDistances)
    .map { (weightedScore, editDistance) -> weightedScore * editDistance }
    .sum()
```

Reply:
Looks good. I'll switch to this.

Reply:
Done. I had to make a small change since I can't use `zip` for the final score calculation because it will ignore scores for words that don't have a corresponding edit distance calculation.

Reply:
> it will ignore scores for words that don't have a corresponding edit distance calculation.

Ah, right. That's what I was worried about with zip.